design/baremetalhost-v1alpha2.md

```diff
@@ -0,0 +1,236 @@
+# BareMetalHost v1alpha2 API Changes
+
+## Status
+
+provisional
+
+## Summary
+
+We need to continue the API version migration for the `BareMetalHost`
+resource type to complete the work described in
+<https://github.com/metal3-io/baremetal-operator/issues/434>. This
+document lists the small API changes to be made and the steps to
+implement the changes.
+
+## Motivation
+
+The `baremetal-operator` repository still relies on a vendored copy of
+the `operator-sdk` project because the `BareMetalHost` CRD includes a
+field with type `float`, which is not formally supported by either
+`operator-sdk` or `kubebuilder`. Changing the field type will require
+a backwards-incompatible API change, so we need to change our API
+version and support upgrading existing users.
+
+### Goals
+
+- Implement a minor API version update to `v1alpha2` for the
+  `BareMetalHost` CRD.
+- Clean up a few issues with the `BareMetalHost` API for which the
+  fixes are not backwards-compatible.
+- Support in-place upgrades of existing `BareMetalHost` resources.
+
+### Non-Goals
+
+- Describe all possible future API changes.
+- Describe a stable `v1` API.
+
+## Proposal
+
+### User Stories
+
+#### Story 1
+
+As a cluster administrator, I want to update the version of metal3 in
+my existing cluster without losing any data or features.
+
+## Design Details
+
+### Choosing a New Type for the Clock Speed Field
+
+Our problematic field holds the clock speed of a CPU in fractional
+megahertz. We assume the fractional portion is noise, so we could
+truncate the value without losing any information.
+
+We have three options for the new type. The advice from the
+`controller-tools` maintainers is to use `resource.Quantity` or
+`int32` for floating point values. We receive the value from
+gophercloud as a string, so we could also simply pass the string
+through unmodified.
+
+The `Quantity` type adds extra characters representing the unit
+conversion, so that a `1 GHz` value would be expressed as `1Gi`, which
+would be a bit ugly for someone trying to consume the value without
+access to the `go` library for parsing the string.
+
+The CPU speed is simply being reported, and not used by the
```

Comment:
There has been interest in the past in using automation to parse specs such as this, compare each to some standard or minimum, and apply labels to hosts based on those comparisons. To support that goal, it will help to keep parsability and comparability in mind in choosing the type. A string would probably not be ideal for that purpose.